### ISSUE TITLE
LDAP integration question

### ISSUE THREAD

**Issue by euanmacinnes** [dbeaver/cloudbeaver#341]
One of the biggest headaches of database admin, and infrastructure is the separate account management of individual apps. Is there a plan to add LDAP support for CloudBeaver, so that we can give consistent account logging information to users, as well as map connections to LDAP roles for centralized security management? 2-4 users isn't much of a problem, but 40-50 users most definitely is, just to give an idea of scale.

**Comment by kseniiaguzeeva** [comment]
Thank you for the idea. We will think what can be done.

**Comment by simonpinnow** [comment]
Hi,
this would be an important feature for us, too. Without it is hard to handle. We want to use it inside of a kubernetes.

Thank you.
Best regards.
Simon

**Comment by Geronium** [comment]
hello, 

some news of the ldap integration ?

Thank you

**Comment by kseniiaguzeeva** [comment]
Could you please give more details and examples how do you use LDAP?  Do you use it for Active Directory only? 
As I see different cases exist and we should understand what way will be the best to integrate it. 
Thank you in advance for your help.

**Comment by Geronium** [comment]
hello,

in my case, we use openldap to authenticate on the applications. Either directly or through the system (which allows it to authenticate via openldap).

you can see an exemple of the entry inside the ldap 

 dn: uid=test,ou=users,idPF=prod,ou=enterprise,c=com
cn: test
gidnumber: 1000
givenname: test
homedirectory: /home/test
loginshell: /bin/bash
mail: test@test.com
objectclass: top
objectclass: person
objectclass: inetOrgPerson
objectclass: posixAccount
sn: test
uid: test
uidnumber: 1000
userpassword: test

and an exemple of application configuration directly 

import ldap
import django_auth_ldap.config
from django_auth_ldap.config import LDAPSearch

AUTH_LDAP_SERVER_URI = "ldap://172.222.1.254:389"
AUTH_LDAP_BIND_DN = "cn=Manager,ou=enteprise,c=com"
AUTH_LDAP_BIND_PASSWORD = "***"
AUTH_LDAP_USER_SEARCH = django_auth_ldap.config.LDAPSearch(
            "ou=users,idPF=prod,ou=enterprise,c=com", ldap.SCOPE_SUBTREE, "uid=%(user)s"
            )

if we pass by the system, we need to authorize the system to authenticate with the ldap, for that i use nslcd and i add the rights in the pam.d configuration for this application 

-(mer. juil. 07 10:22:47)--(myserver:/etc/pam.d)-
[root] # cat application 
#%PAM-1.0
auth      sufficient     pam_ldap.so
auth      requisite      pam_succeed_if.so quiet
auth      required       pam_unix.so nodelay
account   required       pam_unix.so

i don't know what is better, to authenticate by the system (if it's possible) or directly with the ldap, but it's what we need, in my case.

Thank you.

**Comment by dawsongzhao** [comment]
we aslo need ldap to countrol develper access connections

**Comment by kseniiaguzeeva** [comment]
@dawsongzhao thank you for voting. Could you please also give more details and examples how do you use LDAP?

**Comment by simonpinnow** [comment]
Hey, sorry for delay because of holiday.

For us it would be nice to have:

1. a simple ldap bind with the user credentials given during authentication
2. a configured ldap group of which the user must be a member of

configuration:

- ldap-server-hostname
- ldap port
- ssl certificate (optional)
- ldap group

best regards
Simon

**Comment by ThePrez** [comment]
piling on (for better or for worse). I'd like to use an IBM i server for LDAP authentication (so people can log into DBeaver with their IBM i username/pw).

^^ The above is probably non-useful feedback, but I can say that the LDAP support/design in the [gitbucket project](https://github.com/gitbucket/gitbucket) works great for this use case.

**Comment by kseniiaguzeeva** [comment]
It's quite difficult to implement the feature, because different cases exist how LDAP can be used. Thank you for the provided link, we are going to investigate it.
An implimentation is going to be after 21.1 release.

**Comment by Geronium** [comment]
Hello, 

we are at the 22.1.0 version, we have exceeded the 21.1 release and i don't see the ldap fonctionnality. In the millestone 

https://github.com/dbeaver/cloudbeaver/milestones

the ldap fonctionnality is not planned. 

I noticed that in the last release there was the addition of "authentication via nginx" ?

"Users can login to the application via Nginx." 

can we use it to connect via ldap ?

Thank you

**Comment by alexander-skoblikov** [comment]
@Geronium Hello,
 yes, it is potentially possible if you can integrate nginx with ldap and send the correct authorization headers when opening Cloudbeaver. You can read more about required headers and how to set it up on the Cloudbever side on our wiki page - [Reverse proxy authentication
](https://github.com/dbeaver/cloudbeaver/wiki/Reverse-proxy-header-authentication)

**Comment by bsteinweg** [comment]
We use LDAP as a way to connect to Oracle databases, _without_ needing to distribute TNS Names file to all users.  LDAP simply allows easier administration of Oracle DB connections.  Seems like that would be a simple and helpful addition to DBeaver...!

**Comment by jonsbun** [comment]
Would be great to know if this feature planned only for EE or also in Community edition?

**Comment by EvgeniaBzzz** [comment]
@jonsbun 
We plan to add this functionality to all versions.

**Comment by EvgeniaBzzz** [comment]
Thanks everyone for waiting, it's finally done!
LDAP auth is available in CloudBeaver Community edition v.24.1.0 (PRO versions will be released in a week).
Wiki page with configuration will be created soon.

We would appreciate your feedback, any improvements are welcome! 🦫 

-----

**Configuring LDAP in Community edition:**
- open `workspace/.data/.cloudbeaver.runtime.conf` and add configuration to `app.authConfigurations`
(you can also add it to `cloudbeaver/deploy/cloudbeaver/.cloudbeaver.conf` if you prefer to have LDAP in all new workspaces)
```
    "authConfigurations": [
      {
        "id": "ldap",
        "provider": "ldap",
        "displayName": "LDAP",
        "disabled": false,
        "iconURL": "",
        "description": "",
        "parameters": {
          "ldap-host": "your_host",
          "ldap-port": "389",
          "ldap-dn": "ou\your_ou,dc\your_dc1,dc\your_dc2"
        }
      }
    ]
```

   - do not forget to restart a server
- enable LDAP in Administration - Server configuration
- you will see a new LDAP tab on the auth window

**Comment by jonsbun** [comment]
Thank you for this integration. I have a few additional questions:
* What is the correct way to provide LDAP binding account credentials?
* Is it possible to map LDAP users with the CloudBeaver Teams automatically?

**Comment by EvgeniaBzzz** [comment]
@jonsbun 
1. It depends on your LDAP structure.
If you have 1-2 organization units then you can specify them in config (in `ldap-dn`). So you will have to login only with you common name and password.
If you have many organization units then you can specify in config only dc's and login with ou + cn + password.
![image](https://github.com/dbeaver/cloudbeaver/assets/139753579/2b3ecbfa-81fd-4d99-ae72-f84ec094dfc0)

2. LDAP users are created automatically after the first login. There is no user provisioning. So you can't map teams.

**Comment by EvgeniaBzzz** [comment]
Also wiki page is ready, you can find some extra information there.
[LDAP Authentication](https://github.com/dbeaver/cloudbeaver/wiki/LDAP-Authentication)

**Comment by jonsbun** [comment]
> 2. LDAP users are created automatically after the first login. There is no user provisioning. So you can't map teams.

Well, then I think LDAP users mapping with the teams would be very useful. Now the LDAP user first must login and only after that mapping is possible. For example, if a Local users are used, mapping can be achieved in one iteration.

**Comment by EvgeniaBzzz** [comment]
@jonsbun could you, please, create a new issue for that feature?

**Comment by EvgeniaBzzz** [comment]
All pro-versions 24.1.0 with LDAP provider are also released.
I'm closing this issue, but you are always welcome  to leave a comment here or to create a new one 🦫

**Comment by iegurbide** [comment]
I have been giving this a go but I don't think enough information is available to try to integrate with our LDAP system.

Some questions that arise:

- What user is being used to bind to the LDAP server (as ours requires authentication). It is the same user logging in to the specific application i.e no service user is being used? Normally this would be done with a placeholder.

- What field is user name? Generally this should be a configurable unique identifier for the user. Depending on the ldap this attribute could be uid, sAMAccountName, mail, etc. No clue what is being used in this case.

In general, an ldap query looks something like this:

ldapsearch -x  -D "\<bind-user\>" -w "\<bind-password\>"-b "\<base-dn\>" "\<filter\>" -H "\<ldap-host\>:\<ldap-port\>"

I'm assuming mapping from the configuration file is something like this:

- \<base-dn\> --> ldap-dn
-  \<ldap-host\> -->  ldap-host
- \<ldap-port\> --> ldap-port

It would be really helpful to know how the rest of the elements are mapped or if they have even been considered.

**Comment by jwetzell** [comment]
I am in the same boat as @iegurbide it seems there is not enough configurability to support the LDAP setups I see. Gotten numerous other products successfully integrated with LDAP (Harbor, Grafana, Gitea, etc.) without issue.

**Comment by jonsbun** [comment]
> @jonsbun could you, please, create a new issue for that feature?

Feature request: https://github.com/dbeaver/cloudbeaver/issues/2715. I also summarized feedback from the @iegurbide.

**Comment by EvgeniaBzzz** [comment]
New parameters have been added:
`ldap-identifier-attr` - User identifier attribute
`ldap-bind-user` - Bind User DN
`ldap-bind-user-pwd` - Bind User Password
`ldap-filter` - User Filter

Extra information you can find in the [LDAP Authentication](https://github.com/dbeaver/cloudbeaver/wiki/LDAP-Authentication) article

**Comment by LonwoLonwo** [comment]
Hello

We added the ability to set an attribute to be used as the user's login. The new parameter is `ldap-login`. 
The bind user (`ldap-bind-user` and `ldap-bind-user-pwd`) must also be configured to use the attribute.

Now, the parameters look like this:

```
        "parameters": {
          "ldap-host": "ldap.example.com",
          "ldap-port": "1389",
          "ldap-login": "sAMAccountName",
          "ldap-dn": "ou=users,dc=company,dc=com",
          "ldap-identifier-attr": "cn",
          "ldap-bind-user": "cn=serviceAcc,ou=serviceAccounts,dc=company,dc=com",
          "ldap-bind-user-pwd": "serviceAccPas",
          "ldap-filter": ""
        }
```

These changes are found in the latest release, 24.3.2 (06.01.25).

**Comment by DavidRisen** [comment]
Hello,

I have configured the LDAP settings as shown below. The issue is that even with a valid login and password, an error occurs. 
It finds the user, but cannot retrieve their identifier for some reason. In my Active Directory there is no 'userId' field. 
In my case, the user's identifier is stored in sAMAccountName. Perhaps it cannot resolve symbols like CN=Теляков К.Н.? 
But I haven't encountered this problem with any other service. I have already set up more than 10 LDAP services and there were no issues.

For example, let's take Portainer which does not cause any issues and might help to understand the core of the problem:
LDAP Server: personal.aurus.dom:389
Reader DN: CN=Иванов И. И.,OU=otdel10,OU=korpus24,DC=personal,DC=aurus,DC=dom
Password: **************
Base DN: OU=otdel10,OU=korpus24,DC=personal,DC=aurus,DC=dom
Username attribute: sAMAccountName
Filter: (objectclass=user)

The ldap settings in  cloudbeaver.conf:

authConfigurations: [
        {
        "id": "ldap",
        "provider": "ldap",
        "displayName": "LDAP",
        "disabled": false,
        "iconURL": "",
        "description": "",
        "parameters": {
          "ldap-host": "personal.aurus.dom",
          "ldap-port": "389",
          "ldap-login": "sAMAccountName",
          "ldap-dn": "OU=otdel10,OU=korpus 24,DC=personal,DC=aurus,DC=dom",
          "ldap-identifier-attr": "sAMAccountName",
          "ldap-bind-user": "CN=Иванов И. И.,OU=otdel10,OU=korpus 24,DC=personal,DC=aurus,DC=dom",
          "ldap-bind-user-pwd": "************",
          "ldap-filter": "(objectClass=user)"
          }
         }
        ]

Error:
User authentication failed:
LDAP authentication failed: LDAP authentication failed: Failed to determine userId from user DN: CN=Теляков К. Н.,OU=otdel10,OU=korpus 24,DC=personal,DC=aurus,DC=dom
org.jkiss.dbeaver.DBException: Failed to determine userId from user DN: CN=Теляков К. Н.,OU=otdel10,OU=korpus 24,DC=personal,DC=aurus,DC=dom
	at io.cloudbeaver.service.ldap.auth.LdapAuthProvider.findUserNameFromDN(LdapAuthProvider.java:217)
	at io.cloudbeaver.service.ldap.auth.LdapAuthProvider.authenticateLdap(LdapAuthProvider.java:334)
	at io.cloudbeaver.service.ldap.auth.LdapAuthProvider.validateAndLoginUserAccessByUsername(LdapAuthProvider.java:113)
	at io.cloudbeaver.service.ldap.auth.LdapAuthProvider.authExternalUser(LdapAuthProvider.java:78)
	at io.cloudbeaver.service.security.CBEmbeddedSecurityController.authenticate(CBEmbeddedSecurityController.java:1570)
	at io.cloudbeaver.service.auth.impl.WebServiceAuthImpl.authLogin(WebServiceAuthImpl.java:92)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
